plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java

```diff
@@ -128,6 +200,87 @@ public Page getNextPage()
         }
     }
 
+    @Override
+    public void deleteRows(Block rowIds)
+    {
+        RowBlock rows = resolveRowIdBlock(rowIds);
+        posDeleteSink.appendPage(new Page(rows.getPositionCount(), rows.getChildren().toArray(new Block[0])));
+    }
+
+    @Override
+    public void updateRows(Page page, List<Integer> columnValueAndRowIdChannels)
+    {
+        int batchSize = page.getPositionCount();
+        Block rowIdRawBlock = page.getBlock(columnValueAndRowIdChannels.get(columnValueAndRowIdChannels.size() - 1));
+        RowBlock rowIdBlock = resolveRowIdBlock(rowIdRawBlock);
+        Block[] rowIdFieldBlocks = rowIdBlock.getChildren().toArray(new Block[0]);
+        Block[] posDeleteBlocks = new Block[2];
+        posDeleteBlocks[0] = rowIdFieldBlocks[0];
+        posDeleteBlocks[1] = rowIdFieldBlocks[1];
+        posDeleteSink.appendPage(new Page(batchSize, posDeleteBlocks));
+
+        Block[] allTableColumnBlocks = new Block[allTableColumnChannels.length];
+        for (int i = 0; i < allTableColumnChannels.length; i++) {
+            int allTableColumnChannel = allTableColumnChannels[i];
+            if (allTableColumnChannel > -1) {
+                allTableColumnBlocks[i] = rowIdFieldBlocks[allTableColumnChannel + 2];
+            }
+            else {
+                int updateValueChannelIndex = (-1) - allTableColumnChannel;
+                allTableColumnBlocks[i] = page.getBlock(columnValueAndRowIdChannels.get(updateValueChannelIndex));
+            }
+        }
+
+        updateRowSink.appendPage(new Page(batchSize, allTableColumnBlocks));
+    }
+
+    private RowBlock resolveRowIdBlock(Block rowIds)
+    {
+        RowBlock rows;
+        if (rowIds instanceof RowBlock) {
+            rows = (RowBlock) rowIds;
+        }
+        else if (rowIds instanceof DictionaryBlock) {
+            Block dictionary = ((DictionaryBlock) rowIds).getDictionary();
+            if (dictionary instanceof RowBlock) {
+                rows = (RowBlock) dictionary;
```

Comment:
The semantics of `resolveRowIdBlock` are ill-defined. You get "some block, with some values".
the underlying `DictionaryBlock.getDictionary` may contain bogus data, or pretty much anything

remove this method.

you *probably* need it because you didn't the proper class, see the other comment about `Block.getChildren`